### DISCUSSION TITLE
[v5] updateOrCreate problems

### DISCUSSION THREAD

**Discussion by rbartholomay** [adonisjs/core#1565] [Help]
Hi there!

Whats the problem with my code?  On execute i get the error 

_E_INVALID_DATETIME_COLUMN_VALUE: The value for \"MyUser.last_active\" must be an instance of \"luxon.DateTime\""_

Inside model
```javascript
  @column.dateTime({ autoCreate: true })
  public last_active: DateTime
```

```javascript
     const knownUser = await MyUser.updateOrCreate({
        account: account,
      }, {
        account: account,
        ipaddress: ipaddress,
        last_active: DateTime.local(),
      })
```

Bye, René

**Comment by fnoquiq** [comment]
```ts
const knownUser = await MyUser.updateOrCreate({
  account,
}, {
  ipaddress,
  //last_active: DateTime.local(),
})
```

**Reply by rbartholomay** [reply]
that will not work...

What i want is this:

If the User not exists, create a new one.   If the User exists, than update only the updated_at column...  Any idea?

The source below will work only when a user exists. If no user found an errror occurs: "E_INVALID_DATETIME_COLUMN_VALUE: The value for \"WallyUser.updatedAt\" must be an instance of \"luxon.DateTime\""

```javascript
      const knownUser = await WallyUser.updateOrCreate({
        login: login,
      }, {
        login: login,
        ipaddress: ipaddress,
        updatedAt: DateTime.local(),
      })
```

**Reply by thetutlage** [reply]
Nope, you cannot specify different payloads for the `create` and the `update` queries. Also, I am not sure why one would want to only update the updatedAt column vs syncing everything.

You have to do it manually as follows:

```ts
const user = await WallyUser.findBy('login', login)
if (!user) {
  await WallyUser.create()
} else {
  user.updatedAt = DateTime.local()
  await user.save()
}
```

**Reply by rbartholomay** [reply]
ok, this will work...  But when i change my source to i get an error like:

_E_INVALID_DATETIME_COLUMN_VALUE: The value for "MyUser.last_active" must be an instance of "luxon.DateTime""_

```javascript
   const knownUser = await WallyUser.updateOrCreate({
        login: login,
      }, {
        login: login,
        ipaddress: ipaddress,
        lastActivity: DateTime.local()
      })
```

this will work fine, but then it is a string, not a luxon.Datetime. Whats wrong here?

```javascript
   const knownUser = await WallyUser.updateOrCreate({
        login: login,
      }, {
        login: login,
        ipaddress: ipaddress,
        lastActivity: DateTime.local().toFormat('yyyy-MM-dd HH:mm:ss'),
      })
```

**Reply by thetutlage** [reply]
What's changed in this code? It is the same snippet you shared earlier. Look at my comment above

**Comment by rbartholomay** [comment]
i am sorry...  the problem is the luxon Datetime. I try your code like this:

```javascript
   const user = await UserActivity.findBy('login', login)
      if (!user ) {
        await UserActivity.create({ ipaddress: ipaddress, lastActivity: DateTime.local()})
      } else {
        user.ipaddress = ipaddress
        user.lastActivity = DateTime.local()
        await user.save()
      }
```

but get also the error: "E_INVALID_DATETIME_COLUMN_VALUE: The value for \"WallyUserActivity.lastActivity\" must be an instance of \"luxon.DateTime\".

only if i change ``DateTime.local()`` to ``DateTime.local().toFormat('yyyy-MM-dd HH:mm:ss'),`` it will work, but get a typescript error. Inside the model i declare the column like this:

```javascript
...
  @column.dateTime()
  public lastActivity: DateTime
...
```

**Reply by ivanj26** [reply]
Hi @thetutlage  , i am kindly newbie using Adonis v5 and got some error just like yours. I am trying to insert data to model that has column.dateTime by using instance of luxon.Datetime but got this error too . I don't know where 's the problem come from. I think DateTime.fromJSDate should be return the instance of Datetime, right?

 "E_INVALID_DATETIME_COLUMN_VALUE: The value for ..."
Is it something wrong with luxon?

Here's my code:

```javascript
    const startLuxon = DateTime.fromJSDate(startJSDate)
    const endLuxon = DateTime.fromJSDate(endJSDate)
    return creator.related('events')
      .create({
        name,
        description,
        slug,
        thumbnail,
        quota,
        remainingQuota: quota,
        eventCreatorId: creatorId,
        basePrice: price,
        startedAt: startLuxon,
        endAt: endLuxon,
      })
```

nb: pardon my english

**Comment by azinudinachzab** [comment]
Encountering the same problem :/
actually this saving me :

> ```ts
> const user = await WallyUser.findBy('login', login)
> if (!user) {
>   await WallyUser.create()
> } else {
>   user.updatedAt = DateTime.local()
>   await user.save()
> }
> ```

but i knew we waiting for improvement, thanks.

**Comment by hariprasath-yadav** [comment]
simply downgrade you lucid package with

> npm i --save @adonisjs/lucid@^8.4.1

and it solved my many other problems too

**Reply by azinudinachzab** [reply]
this solution works on me. Thanks

**Comment by FeMaffezzolli** [comment]
I'm facing the same issue when trying to create (Model.create({ ... })).

Exception: E_INVALID_DATETIME_COLUMN_VALUE.

Even if I pass a DateTime.local() to the field I encounter this error.

Here's my package.json

```
"devDependencies": {
    "@adonisjs/assembler": "^3.0.0",
    "@types/lodash": "^4.14.168",
    "@types/qs": "^6.9.6",
    "@types/uuid": "^8.3.0",
    "adonis-preset-ts": "^2.0.0",
    "eslint": "^7.20.0",
    "eslint-config-prettier": "^7.2.0",
    "eslint-plugin-adonis": "^1.2.0",
    "eslint-plugin-prettier": "^3.3.1",
    "pino-pretty": "^4.5.0",
    "prettier": "^2.2.1",
    "typescript": "^4.1.5",
    "youch": "^2.1.1",
    "youch-terminal": "^1.0.1"
  },
  "dependencies": {
    "@adonisjs/auth": "^5.1.1",
    "@adonisjs/core": "^5.0.4-preview-rc",
    "@adonisjs/lucid": "^10.0.0",
    "@adonisjs/mail": "^5.2.3",
    "@adonisjs/repl": "^1.0.0",
    "@adonisjs/view": "^3.1.0",
    "aws-sdk": "^2.843.0",
    "axios": "^0.21.1",
    "lodash": "^4.17.21",
    "luxon": "^1.26.0",
    "pg": "^8.5.1",
    "phc-argon2": "^1.1.0",
    "proxy-addr": "^2.0.6",
    "qs": "^6.9.6",
    "reflect-metadata": "^0.1.13",
    "source-map-support": "^0.5.19",
    "uuid": "^8.3.2"
  }
```

Is there any solution available other than downgrading lucid?


### SOLUTION

The problem is Luxon. Downgrading from 1.26.0 to 1.25.0 works fine.

**Reply by thetutlage** [reply]
Fixed in https://github.com/adonisjs/lucid/commit/f283738a60ed0441af0e3ec3f4800572c42efe35
